libkbfs/modes.go

```diff
@@ -0,0 +1,310 @@
+}
+
+func (mm modeMinimal) DirtyBlockCacheEnabled() bool {
+	// No blocks will be dirtied in minimal mode, so don't bother with
+	// the dirty block cache.
+	return false
+}
+
+func (mm modeMinimal) BackgroundFlushesEnabled() bool {
+	// Don't do background flushes when in minimal mode, since there
+	// shouldn't be any data writes.
+	return false
+}
+
+func (mm modeMinimal) MetricsEnabled() bool {
+	return false
+}
+
+func (mm modeMinimal) ConflictResolutionEnabled() bool {
+	// No need to run CR if there won't be any data writes on this
+	// device.  (There may still be rekey writes, but we don't allow
+	// conflicts to happen in that case.)
+	return false
+}
+
+func (mm modeMinimal) BlockManagementEnabled() bool {
+	// If this device is in minimal mode and won't be doing any data
+	// writes, no need deal with block-level cleanup operations.
+	// TODO: in the future it might still be useful to have
+	// e.g. mobile devices doing QR.
+	return false
+}
+
+func (mm modeMinimal) QuotaReclamationEnabled() bool {
+	return false
+}
+
+func (mm modeMinimal) NodeCacheEnabled() bool {
+	// If we're in minimal mode, let the node cache remain nil to
+	// ensure that the user doesn't try any data reads or writes.
+	return false
+}
+
+func (mm modeMinimal) TLFUpdatesEnabled() bool {
+	return true
+}
+
+func (mm modeMinimal) KBFSServiceEnabled() bool {
+	return false
+}
+
+func (mm modeMinimal) JournalEnabled() bool {
+	return false
+}
+
+func (mm modeMinimal) UnmergedTLFsEnabled() bool {
+	// Writes aren't allowed, so unmerged TLFs on this device
+	// shouldn't be possible.
+	return false
+}
+
+func (mm modeMinimal) ServiceKeepaliveEnabled() bool {
+	return false
+}
+
+func (mm modeMinimal) TLFEditHistoryEnabled() bool {
+	return false
+}
+
+func (mm modeMinimal) ClientType() keybase1.ClientType {
+	return keybase1.ClientType_KBFS
+}
+
+// Single op mode:
+
+type modeSingleOp struct {
+	InitMode
+}
+
+func (mso modeSingleOp) Type() InitModeType {
+	return InitSingleOp
+}
+
+func (mso modeSingleOp) RekeyWorkers() int {
+	// Just block all rekeys and don't bother cleaning up requests
+	// since the process is short lived anyway.
+	return 0
+}
+
+func (mso modeSingleOp) QuotaReclamationEnabled() bool {
+	return false
+}
+
+func (mso modeSingleOp) TLFUpdatesEnabled() bool {
+	return false
+}
+
+func (mso modeSingleOp) KBFSServiceEnabled() bool {
+	return false
+}
+
+func (mso modeSingleOp) UnmergedTLFsEnabled() bool {
+	// There's basically no way for a TLF to start off as unmerged
+	// since single-ops should be using a fresh journal.
+	return false
+}
+
+func (mso modeSingleOp) TLFEditHistoryEnabled() bool {
+	return false
+}
+
+func (mso modeSingleOp) ClientType() keybase1.ClientType {
+	return keybase1.ClientType_NONE
+}
+
+// Constrained mode:
+
+type modeConstrained struct {
+	InitMode
+}
+
+func (mc modeConstrained) Type() InitModeType {
+	return InitConstrained
+}
+
+func (mc modeConstrained) BlockWorkers() int {
+	return 1
```

Comment:
I know this is from @jzila 's temporary PR, but now that we are formalize it, do we maybe want to use a larger number here? Especially considering mobile connections probably have higher latency.

Reply:
We can start with 1 and change it later to optimize.